### PR TITLE
Add experimental stacked RPA attention kernel

### PR DESCRIPTION
**Pull request by rupengliu-meta**
# Description
New Stacked rpa kernel that support up to 1M context length and use seq_alone_lane kv layout by default.
Upload an initial version as discussed, more optimization will be added later
initial results compared with batched rpa on decode performance. Batched rpa is still faster on short context, while stack is faster on long context. 
截屏2026-07-23 上午10 13 10

Start with a short description of what the PR does and how this is a change from
the past.

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a Github issue, please include a link, e.g.,:
FIXES: #123456

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/kernels/experimental/stacked_rpa/flash_attention.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+                                     axis=0)
+                return jnp.stack([mask_tq for _ in range(k_heads)], axis=0)
+
+            q_idx_b = q_offset_b.astype(int_ty) + processed_q_len[b_idx]
+            eff_kv_len_b = effective_kv_len[b_idx]
+            mask = q_idx_b < eff_kv_len_b
+            mask = jnp.logical_and(mask, q_idx_b >= kv_idx_b)
+
+            if (sliding_window := cfgs.model.sliding_window) is not None:
+                mask = jnp.logical_and(mask, q_idx_b
```
Comment (qizzzh) [inline tpu_inference/kernels/experimental/stacked_rpa/flash_attention.py:167]: `int_ty = cfgs.serve.int_ty` (line 136) is caller-selectable. When callers pick `jnp.int16` (a natural optimization for `max_model_len ≤ 32767` on v6e/v7x), the addition `kv_idx_b + sliding_window` at lines 167–168 stays int16 and can exceed the int16 range → wraps negative.

**Failure:** Gemma-style config with `max_model_len=32000`, `sliding_window=4096`, `int_ty=int16` on v6e/v7x. For `kv_idx_b` near `max_model_len`, `kv_idx_b + sliding_window > 32767` wraps negative — the predicate `q_idx_b < kv_idx_b + sliding_window` flips False for valid keys inside the window. Combined with the causal mask above, valid tokens get blanked → wrong logits.

**Fix:** cast `kv_idx_b` (or the sum) to int32 before the compare, or auto-widen `int_ty` when `max_model_len + sliding_window > INT16_MAX`.

File: tpu_inference/kernels/experimental/stacked_rpa/wrapper.py [hunk 2]
```diff
@@ -0,0 +1,1374 @@
...
+
+    ``prepacked_new_kv_hbm``: optional post-RoPE K/V already packed in the
+    SEQ_ALONG_LANE ``new_kv_hbm`` layout. When provided, ``prepare_inputs`` skips
+    the K/V concat+pad+transpose and only prepares Q.
+
+    Args:
+        queries: [max_num_tokens, num_q_heads, head_dim]. Output of q projection.
+        keys: [max_num_tokens, num_kv_heads, head_dim]. Output of k projection.
+        values: [max_num_tokens, num_kv_heads, head_dim]. Output of v projection.
+        kv_cache: [num_pages, page_size, cdiv(num_kv_heads * 2, kv_packing),
```
Comment (qizzzh) [inline tpu_inference/kernels/experimental/stacked_rpa/wrapper.py:1195]: Docstring documents the batched-RPA 5D HEAD_ALONG_SUBLANE layout:
```
kv_cache: [num_pages, page_size, cdiv(num_kv_heads * 2, kv_packing), kv_packing, head_dim]
```
But `get_kv_cache_shape`, `_validate_prepacked_inputs`, and `configs.validate_inputs` all enforce the 4D SEQ_ALONG_LANE layout `(num_pages, num_kv_heads*2, aligned_head_dim, page_size)`.

**Failure:** User reads the docstring, allocates the 5D shape. `page_size = kv_cache.shape[3]` (in `_resolve_attn_static`) silently becomes `kv_packing (=2)`; downstream VMEM math and `_clamp_bkv_to_page` corrupt on `page_size=2`; validation finally raises a shape complaint naming `num_kv_heads*2` — the user cannot correlate the error back to the docstring they followed.

**Fix:** rewrite this docstring to match `get_kv_cache_shape`, or (better) point the docstring at `get_kv_cache_shape` and remove the divergent per-arg description.

**Review by qizzzh — APPROVED** [review]

File: tpu_inference/kernels/experimental/stacked_rpa/bref_override.py [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+from tpu_inference.kernels.experimental.stacked_rpa import configs, schedule
+
+# `pltpu.BufferType` compatibility shim. The JAX that ships with the build image
+# (jax==0.9.2) does not expose `pltpu.BufferType`; the real `buffer_type` value is
+# passed through from `emit_pipeline` to `BufferedRef.create`, so the annotations
+# and asserts below are advisory only. Provide a permissive stand-in whose enum
+# members compare equal to anything, turning those asserts into no-ops (matching
+# the upstream base, which kept them commented out). Remove once the image ships
+# a JAX that exposes `pltpu.BufferType`.
+if not hasattr(pltpu, "BufferType"):
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/stacked_rpa/bref_override.py:31]: We can just remove the type checking. That how we did before.

File: tpu_inference/kernels/experimental/stacked_rpa/envvars.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+
+def stacked_rpa_bkv() -> int:
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/stacked_rpa/envvars.py:18]: Please always make kernel stateless and not reference any global variables. Access it via arguments.

File: tpu_inference/kernels/experimental/stacked_rpa/flash_attention.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+from tpu_inference.kernels.experimental.stacked_rpa import configs, utils
+
+# ---------------------------------------------------------------------------
+# `pltpu.einshape` compatibility shim. Older JAX nightlies (e.g. older
+# deployed images on libtpu==0.0.37.dev) don't ship einshape.
+# Implement a focused fallback covering the merge/split patterns used in this
+# file. Remove once the deployed image is rebuilt with jax==0.9.2.
+# ---------------------------------------------------------------------------
+if not hasattr(pltpu, "einshape"):
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/stacked_rpa/flash_attention.py:28]: I'm seeing a lot of this pattern. Why? It's not like this function provides really necessary functionality. You are adding around 100 lines of code just to avoid reshape?

File: tpu_inference/kernels/experimental/stacked_rpa/flash_attention.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+            return x.reshape(new_shape)
+
+        # Pure permutation
+        in_axis = {t: i for i, t in enumerate(in_toks)}
+        return jnp.transpose(x, [in_axis[t] for t in out_toks])
+
+    pltpu.einshape = _einshape_compat
+
+
+@jax.named_scope("flash_qk_softmax")
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/stacked_rpa/flash_attention.py:87]: This can potentially induce optimization barrier.

File: tpu_inference/kernels/experimental/stacked_rpa/flash_attention.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+            mask = q_idx_b < eff_kv_len_b
+            mask = jnp.logical_and(mask, q_idx_b >= kv_idx_b)
+
+            if (sliding_window := cfgs.model.sliding_window) is not None:
+                mask = jnp.logical_and(mask, q_idx_b
+                                       < kv_idx_b + sliding_window)
+            return mask
+
+        if skip_mask is not None:
+            qk_masked_b = lax.cond(
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/stacked_rpa/flash_attention.py:172]: This introduced optimization barrier.

File: tpu_inference/kernels/experimental/stacked_rpa/kernel.py [hunk 8]
```diff
@@ -0,0 +1,1092 @@
...
+#
+# Upstream introduced `pl.with_scoped` (a decorator form of `pl.run_scoped`) in
+# the JAX nightly that ships with the current tpu-inference image. Older JAX
+# nightlies (e.g. older deployed images on libtpu==0.0.37.dev)
+# only have `pl.run_scoped`. Synthesize `with_scoped` from `run_scoped` on
+# those older versions so the vendored kernel runs on both.
+# Remove this shim once the deployed image is rebuilt with jax==0.9.2 (which
+# ships `pl.with_scoped` natively).
+# ---------------------------------------------------------------------------
+if not hasattr(pl, "with_scoped"):
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/stacked_rpa/kernel.py:41]: Again: you are adding this just to avoid passing scratch memory in Pallas call.

**Comment by rupengliu-meta** [comment]
cloing the PR for now

**Review by qizzzh — APPROVED** [review]
Approving to unblock. Once this is merged, I'll update https://github.com/vllm-project/tpu-inference/pull/3330 to only show the diff.

**Review by kyuyeunk — APPROVED** [review]

**Comment by kyuyeunk** [comment]
approving as it is in experimental folder without potential impact for other workloads.
